### PR TITLE
Expose child routes on currentRoute object; add property for a CSS class on inactive routes

### PR DESCRIPTION
**Pull request by tmcdos**
It is not easy to get access to child routes when you need to render nested menus with `<router-link>`
Publishing **children** on the **$route.currentRoute** solves a real problem. The **parent** property is required so that the route can see its siblings. The **title** property is optional - it is useful instead of **name** when you want to name a route having a default child. I personally use **name** in `beforeRouteEnter/beforeRouteUpdate` to update the **document.title** in browser - and I need it on all routes (even with default child), so I had to add **title** property.

This closes issue #1149

There are non-trivial situations where playing with CSS selectors becomes too inconvenient. Consider the following markup:
```vue
    <div>
      <div class="tabs_account">
        <div class="tab_spacer">&nbsp;</div>
        <template v-for="page in $router.currentRoute.parent.children" v-if="page.can_show">
          <router-link tag="div" active-class="tab_active" inactive-class="tab" v-bind:to="page.path">{{ page.title }}</router-link>
          <div class="tab_spacer">&nbsp;</div>
        </template>
        <div class="tab_fill tab_spacer">&nbsp;</div>
      </div>
      <router-view></router-view>
    </div>
```
You can achieve the same effect like this
```vue
<router-link :class="{'router-link-active': $route.fullPath ==='/' || $route.fullPath === '/a', 'router-link-inactive': $route.fullPath !=='/' && $route.fullPath !== '/a'}" to="/a">
Text
</router-link>
```
but it looks hacky (you have to keep the PATH strings in sync between HTML and JavaScript code) while the proposed solution feels more elegant.

### CODE REVIEW

**Comment by posva** [comment]
Hey thanks for the PR 🙂 
If you want to add the feature, please, only add **that** feature, so don't add the `title` feature which is a personal thing. Also, please, don't add the `// TMCDOS` at the end of every line you added
Remember to add a test case for children

**Comment by tmcdos** [comment]
Okay, I followed your recommendations:

- removed the `// TMCDOS` marks
- removed the custom **title** property
- added a unit test for **parent** and **children[]** properties
- added a Selenium test for **inactive-class** and **exact-inactive-class**
- fixed some settings to allow headless browser testing with ChromeDriver

Hopefully the PR is now acceptable.

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "lint": "eslint src examples",
     "test": "npm run lint && flow check && npm run test:unit && npm run test:e2e && npm run test:types",
     "test:unit": "jasmine JASMINE_CONFIG_PATH=test/unit/jasmine.json",
-    "test:e2e": "node test/e2e/runner.js",
+    "test:e2e": "xvfb-run node test/e2e/runner.js && rm -rf ./xvfb-run.* && rm -rf ./.com.google.* && rm -rf ./.org.chromium.*",
```
Comment (posva) [inline package.json:31]: Please let it as it was

File: test/e2e/nightwatch.config.js [hunk 2]
```diff
@@ -33,7 +33,10 @@ module.exports = {
       'desiredCapabilities': {
         'browserName': 'chrome',
         'javascriptEnabled': true,
-        'acceptSslCerts': true
+        'acceptSslCerts': true,
```
Comment (posva) [inline test/e2e/nightwatch.config.js:36]: Can you remove this as well please

File: test/unit/specs/children.spec.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import Vue from 'vue'
+import VueRouter from '../../../src/index'
+
+Vue.use(VueRouter)
+
+describe('currentRoute', () => {
+  describe('children[]', () => {
```
Comment (posva) [inline test/unit/specs/children.spec.js:7]: I don't think a describe is necessary here. a test with _should have children_ and another with _should have parent.children_ is enough

**Review by posva — CHANGES_REQUESTED** [review]
I thought you were going to remove the other extra feature as well 😅 . Can you remove the inactive class feature from this PR, please?
Also, could you add the typings to the `flow/declarations.js` file, please?

**Comment by Frondor** [comment]
I believe this is overkill, child routes should never be exposed in `currentRoute` object. A new method like `this.$route.getChildren()` should be better in my opinion.

**Comment by tmcdos** [comment]
Child routes are exposed in every **route** object - just like the **parent** property. Why do you think a method is better than a property ?

**Comment by Frondor** [comment]
Actually, no. They aren't.
The parent route is just one referenced object, and if you are exposing children for those route objects, you're wasting a lot of memory for something that most people don't need. That's why **in my opinion** it should be called/built on demand instead.

**Comment by tmcdos** [comment]
Forgive me my profanity - perhaps using the original JavaScript objects as defined when instantiating the router instead of partially cloning them and exposing these partial clones would solve the problem with memory waste ?

**Comment by kanatkubash** [comment]
Are not js objects just referenced not copied, it shouldnt be wasting memory.

**Comment by posva** [comment]
Closing in favour of #1991

**Pull request closed without merging**